internal/insight/insight.go

```diff
@@ -65,17 +66,37 @@ func (in *Insight) Fetch(ctx context.Context, dbiResourceId string, dur time.Dur
 		chunks = append(chunks, metrics[i:end])
 	}
 
-	// TODO: concurrent
+	childContext, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	samples := map[string]Samples{}
-	for _, chunk := range chunks {
-		set, err := in.fetch(ctx, dbiResourceId, dur, chunk...)
-		if err != nil {
-			return nil, err
-		}
+	var wg sync.WaitGroup
+	var err error
 
-		for k, v := range set {
-			samples[k] = v
-		}
+	for _, chunk := range chunks {
+		chunk := chunk
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			set, e := in.fetch(childContext, dbiResourceId, dur, chunk...)
+			if e != nil {
+				if e != context.Canceled {
+					cancel()
+					err = e
+				}
+				return
+			}
+
+			for k, v := range set {
+				samples[k] = v
```

Comment:
Please note that maps in Golang are not "thread" safe. https://go.dev/blog/maps#concurrency

We have to either protect writes to map with mutex (simples) or use another cannel to emit `set` into main routine (hardest).

Reply:
I totally forgot that. 👍